src/lib/ffi/ffi_cipher.cpp

```diff
@@ -132,81 +140,117 @@ int botan_cipher_start(botan_cipher_t cipher_obj, const uint8_t* nonce, size_t n
+      if(final_input) {
+         // If the final flag is set for the first time, we need to process the
+         // remaining input data and then finalize the cipher object.
+         if(!was_finished_before) {
+            *input_consumed = input_size;
+            mbuf.resize(input_size);
+            copy_mem(mbuf, std::span(input, input_size));
+
+            try {
+               cipher.finish(mbuf);
+            } catch(Invalid_Authentication_Tag&) {
+               return BOTAN_FFI_ERROR_BAD_MAC;
+            }
+         }
 
-         input_size -= ud;
-         input += ud;
-         taken += ud;
+         // At this point, the cipher object is finalized (potentially in a
+         // previous invocation) and we can copy the final output to the caller.
+         *output_written = mbuf.size();
 
-         if(bytes_produced > 0) {
-            copy_mem(output, mbuf.data(), bytes_produced);
-            output_size -= bytes_produced;
-            output += bytes_produced;
-            written += bytes_produced;
+         // Not enough space to copy the final output out to the caller.
+         // Inform them how much space we need for a successful operation.
+         if(output_size < mbuf.size()) {
+            // This is the only place where mbuf is not cleared before returning.
+            clean_buffer.disengage();
+            return BOTAN_FFI_ERROR_INSUFFICIENT_BUFFER_SPACE;
          }
-      }
 
-      *output_written = written;
-      *input_consumed = taken;
+         // Copy the final output to the caller, mbuf is cleared afterwards.
+         copy_mem(std::span(output, mbuf.size()), mbuf);
+      } else {
+         // Process data in a streamed fashion without finalizing. No data is
+         // ever retained in the cipher object's internal buffer. If we run out
+         // of either input data or output capacity, we stop and report that not
+         // all bytes were processed via *output_written and *input_consumed.
+
+         BufferSlicer in({input, input_size});
+         BufferStuffer out({output, output_size});
+
+         // Helper function to do blockwise processing of data.
+         auto blockwise_update = [&](const size_t granularity) {
+            if(granularity == 0) {
+               return;
+            }
+
+            const size_t expected_output_per_iteration = cipher.requires_entire_message() ? 0 : granularity;
+            mbuf.resize(granularity);
+
+            while(in.remaining() >= granularity && out.remaining_capacity() >= expected_output_per_iteration) {
+               copy_mem(mbuf, in.take(granularity));
+               const auto written_bytes = cipher.process(mbuf);
+               BOTAN_DEBUG_ASSERT(written_bytes == expected_output_per_iteration);
+               if(written_bytes > 0) {
+                  BOTAN_ASSERT_NOMSG(written_bytes <= granularity);
+                  copy_mem(out.next(written_bytes), std::span(mbuf).first(written_bytes));
+               }
+            }
```

Comment:
@randombit There's more room for optimization here. We could detect that the user passed the same buffer for input and output and omit the copy to the internal `mbuf` in that case.